### PR TITLE
Native SMC: batched draws, autobatched potentials, llamppl removed

### PR DESCRIPTION
**Pull request by shepardxia**
Removes llamppl, owns SMC natively, and batches everything a step does concurrently.

## SMC

`SMC` validates and configures, `smc_standard` is the algorithm, `SequenceModel` is the particle, `Sequences` carries the results.

```python
particles = await smc_standard(model, n_particles, ess_threshold=0.5)
```

`smc_standard` clones the template, gathers `step()` over live particles, records, tests ESS, and resamples by clone. One call is one SMC problem. Run several at once with `asyncio.gather` and their potential calls meet in the same batch windows.

`SequenceModel.step()` owns step semantics: draw, score, critic twist and settle, terminate. Samplers expose `sample(context, draw=None) -> (token, logw, logp)` and nothing else.

`token_sampler.smc(...)` keeps its signature. Breaking changes: `SequenceModel` and `Importance` leave the public `sampler` namespace, `flatten_units` moves to `genlm.control.util`, and `llamppl` is dropped as a dependency.

## Batched draws

Concurrent draws across particles, across concurrent SMC runs, and inside unit-sampler loops meet in a per-event-loop window and resolve as one stacked reduction per `(backend, device, vocab)` group, off a single host readback per cohort. Rows stay on the backend's device until that readback.

`draw_from(..., target=...)` makes an entry an importance draw, sampled from the row and weighed under the target, served in the same readback.

`set_draw_method` selects the categorical sampler: `gumbel_max` (default), `multinomial`, `inverse_cdf`. All three draw the same distribution and differ in RNG consumption, which makes `inverse_cdf` the one to reach for in common-random-number experiments.

## Autobatched potentials

`SMC`, `DirectTokenSampler` and `AWRS` wrap their potential seats in `AutoBatchedPotential` at construction, so concurrent per-particle asks reach the potential as one batched call. Pass `autobatch=False` to opt out. `TrieSetSampler` wraps only `iter_potential`, since the trie walk asks `item_potential` sequentially. `autobatched()` is one memoized door whose memo lives on the potential instance, so shared potentials share a window and the wrapper dies with its potential.

The old `AsyncBatchLoop` background task is deleted.

## Reweighting

`Tempered` and `Normalized` are new:

```python
p ** beta                      # scales every log weight; -inf survives any beta
p.normalize()                  # renormalizes each next-token row
(p ** (1 / tau)).normalize()   # p at temperature tau
```

## Vocabulary-level sharing

`Potential.build_tables` with `tables=` skips the O(V) `__init__`. `Coerced.build_trie` with `trie=` shares one symbol trie across coercions over a vocabulary. `live_logws` lets a potential enumerate its live tokens, so `batch_logw_next` scatters a whole batch into one `alloc_rows` block; overriding `alloc_rows` places that block on a device. `coerce(..., homomorphic=)` declares whether `f` distributes over concatenation, which is the condition for taking the shared-prefix trie lane.

## Testing

Behavioral tests only. New coverage for the new machinery: `tests/test_draw_window.py`, `tests/sampler/test_smc_model.py`, `tests/potential/test_autobatch.py`, `tests/potential/test_reweight.py`. The engine serving contract is asserted by the backend's GPU probe at row grain rather than through SMC statistics. `pytest-xdist` is added for parallel runs.

CI installs genlm-backend from `requirements-ci.txt`, which points at the co-developed branch. That file goes away when the backend releases and the floor in `pyproject.toml` moves up.

Lands with [genlm-backend #74](https://github.com/genlm/genlm-backend/pull/74). Each side needs the other's branch.

### CODE REVIEW

File: genlm/control/sampler/smc.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+                record.add_smc_step(particles)
+
+        ancestor_indices = None
+        W = np.array([p.weight for p in particles])
+        if np.all(W == -np.inf):
+            continue
+        w_sum = logsumexp(W)
+        nw = W - w_sum
+        with np.errstate(divide="ignore"):
+            if -logsumexp(nw * 2) >= np.log(ess_threshold) + np.log(n_particles):
```
Comment (samuki) [inline genlm/control/sampler/smc.py]: In llamppl we were resampling `if -logsumexp(normalized_weights * 2) < np.log(ess_threshold) + np.log(n_particles)`. I think reversing the condition could break if the probabilities contain NaN. Are we catching this case earlier, or do we need to add a case for that here?

File: genlm/control/sampler/smc.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+                # `finish` took the twist back; at termination the critic's score
+                # is real weight.
+                self.score(twist_amt)
+
+    def _append(self, unit):
+        """Extend the context by one drawn unit.
+
+        A multi-token unit ending in EOS is split, so that ``context[-1] is EOS``
+        whenever the sequence is terminal. ``terminate_when`` appends EOS in the
+        step that satisfied it and carries no weight correction: the stop
```
Comment (samuki) [inline genlm/control/sampler/smc.py:160]: Should we also apply a weight correction here?

File: genlm/control/potential/autobatch.py [hunk 3]
```diff
@@ -1,57 +1,95 @@
...
+                if isinstance(results, LazyWeights):
+                    results = [
+                        results.spawn(results.weights[i])
+                        for i in range(len(requests))
+                    ]
+                assert len(results) == len(requests)
+                for (_, future), result in zip(requests, results):
+                    if not future.done():
+                        future.set_result(result)
+            except Exception as exc:
```
Comment (samuki) [inline genlm/control/potential/autobatch.py:78]: Should this except `BaseException`?

File: genlm/control/util.py [hunk 4]
```diff
@@ -287,49 +332,223 @@ def load_async_trie(V, backend=None, **kwargs):
...
 
-def fast_sample_logprobs(logprobs: np.ndarray, size: int = 1) -> np.ndarray:
-    """Sample indices from an array of log probabilities using the Gumbel-max trick.
+# --- token-picker family ---
+# Each maps a log-weight tensor -> drawn index over dim=-1 (1-D draw or batched [N, V]).
+
+
+def gumbel_max(logps):
+    """Argmax of `logps + Gumbel noise`; the default picker."""
+    g = -torch.log(-torch.log(torch.rand_like(logps)))
```
Comment (samuki) [inline genlm/control/util.py:349]: These default to fp32. Should we enforce fp64 here?

File: pyproject.toml [hunk 5]
```diff
@@ -18,7 +18,6 @@ dependencies = [
     "genlm-grammar>=0.2.0",
     "genlm-backend>=0.2.0",
```
Comment (samuki) [inline pyproject.toml:19]: We'll have to update this before the merge.

File: genlm/control/util.py [hunk 6]
```diff
@@ -287,49 +332,223 @@ def load_async_trie(V, backend=None, **kwargs):
...
+    flattened = []
+    for item in context:
+        if isinstance(item, list):
+            flattened.extend(flatten_units(item))
+        else:
+            flattened.append(item)
+    return flattened
+
+
+DRAW_METHODS = {
```
Comment (samuki) [inline genlm/control/util.py:382]: I think we should document the new handling of seeds and RNGs since it changed a lot compared to main.

File: tests/potential/test_duplicate_byte_strings.py [hunk 7]
```diff
@@ -139,32 +138,6 @@ async def test_duplicate_tokens_get_independent_weights(llm):
...
-        llm.token_maps.eos_idxs
-    )
-
-
-# ---------------------------------------------------------------------------
-# Tokenize / encode / decode roundtrip
-# ---------------------------------------------------------------------------
-
-
-def test_tokenize_roundtrip(llm):
```
Comment (samuki) [inline tests/potential/test_duplicate_byte_strings.py:159]: Why was the tokenize roundtrip test deleted?

**Review by samuki — CHANGES_REQUESTED** [review]
The refactor mostly looks good on a first glance. I put in a few questions and will run additional tests on the domains in `genlm-eval`.

File: genlm/control/sampler/token.py [hunk 8]
```diff
@@ -319,18 +292,48 @@ def __init__(
...
+        if c is None or c.device != device:
+            c = self._valid_idxs_cache = torch.as_tensor(
+                self.valid_idxs, dtype=torch.int64, device=device
+            )
+        return c
+
+    def _gen(self, device):
+        g = self._gen_cache
+        if g is None or g.device != torch.device(device):
+            g = self._gen_cache = torch.Generator(device=device)
```
Comment (vicky-xef) [inline genlm/control/sampler/token.py:319]: Is it intended that AWRS(..., seed=None) is deterministic? manual_seed is only called when a seed is given, so every unseeded sampler draws the same Gumbel key stream in `_make_keys`, right? How we want seed=None to behave. We can also add some tests for the `None` path (two unseeded samplers differ or match, two
same-seed samplers match)?

File: genlm/control/util.py [hunk 9]
```diff
@@ -287,49 +332,266 @@ def load_async_trie(V, backend=None, **kwargs):
...
+    _picker = DRAW_METHODS[method] if isinstance(method, str) else method
+
+
+# Batching counters: (site, cohort_size) -> count. Sites are "draw" (one entry per
+# stacked group per flush) and "autobatch" (one entry per batch call, see
+# potential/autobatch.py). Read and clear via `take_batch_stats`.
+batch_stats = Counter()
+
+
+def take_batch_stats():
```
Comment (samuki) [inline genlm/control/util.py:409]: `take_batch_stats()` makes a snapshot of the counter by rebinding the global. 
`potential/autobatch.py` imports the Counter object `from genlm.control.util import batch_stats`,i.e., after the first call it keeps incrementing the old Counter and the ("autobatch", method, n) entries never reach a caller. I think having `autobatch.py` read `util.batch_stats` at the call time would fix it.

File: genlm/control/util.py
Comment (vicky-xef) [inline genlm/control/util.py:50]: is weights a np.ndarray, isn't it a torch.Tensor, on the GPU under vLLM?

Comment (vicky-xef) [inline genlm/control/util.py:61]: same comment for type of weights here as in line 50.

Comment (vicky-xef) [inline genlm/control/util.py:197]: same comment for type of weights here as in line 50.

**Review by vicky-xef — COMMENTED** [review]
We should also check `docs/` folder and `README.md` files before merging
